sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ScalaUDF.scala

```diff
@@ -124,10 +124,10 @@ case class ScalaUDF(
     val toRow = enc.createSerializer().asInstanceOf[Any => Any]
     if (enc.isSerializedAsStructForTopLevel) {
       value: Any =>
-        if (value == null) null else toRow(value).asInstanceOf[InternalRow]
+        if (value == null) null else toRow(value).asInstanceOf[InternalRow].copy()
     } else {
       value: Any =>
-        if (value == null) null else toRow(value).asInstanceOf[InternalRow].get(0, dataType)
+        if (value == null) null else toRow(value).asInstanceOf[InternalRow].copy().get(0, dataType)
```

Comment:
Why do we need `copy` now after #28979? IIUC blocking plans (e.g., aggregates and sorts) copy results in their plan side if necessary.

Reply:
The source of the issue is that `ScalaUDF` is called multiple times from `ArrayTransform` (or any other higher order function). But each time the same `resultConverter` uses the same expression encoder to serialize the UDF's result into an `InternalRow` (https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ScalaUDF.scala#L124) but the encoders are are allowed to return the same instance of `InternalRow` (https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala#L382-L389) and the caller needs to make sure the row is copied if needed. I think https://github.com/apache/spark/pull/28979 simply just forgot to add `.copy()` after these invocations.

Reply:
If so, it looks like this is a bug of the higher order function. Why don't we copy it in the higher order function side?
For example,
```
--- a/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/higherOrderFunctions.scala
+++ b/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/higherOrderFunctions.scala
@@ -279,7 +279,8 @@ case class ArrayTransform(
       if (indexVar.isDefined) {
         indexVar.get.value.set(i)
       }
-      result.update(i, f.eval(inputRow))
+      val v = InternalRow.copyValue(f.eval(inputRow))
+      result.update(i, v)
       i += 1
     }
     result
```
I think the current fix can cause performance regression in case of `ScalaUDF + non higher order functions`.

Reply:
> I think the current fix can cause performance regression in case of ScalaUDF + non higher order functions.
Well, yes, in case of *typed* `ScalaUDF`s my change can cause regression. But your suggested change can cause regression in case of other `UDF`s + `higher order functions` which is admittedly rarer than typed `ScalaUDF` + `non higher order functions`.
So I'm not against fixing this in all affected higher order functions, but I wouldn't expect `ScalaUDF.eval()` to return with something that needs to be copied first to make sure another invocation doesn't change it.

Reply:
I'm not 100% sure that we can assume a previous result is valid if an expr invokes its child `eval` twice. What I'm worried about is that the same bug can happen if the other expressions have an internal buffer just like `ScalaUDF`.

Reply:
I've reverted the ScalaUDF change in https://github.com/apache/spark/pull/31955/commits/6feab1aa4bc059f57fde77d71fd2f95ea87b31c0 and fixed affected higher order functions here: https://github.com/apache/spark/pull/31955/commits/3d22e3bf3cb570302115209132fc9f55c731fa63 tests are in https://github.com/apache/spark/pull/31955/commits/d00e70596e1b679d2331375629249905e9c75395

Reply:
cc: @hvanhovell @cloud-fan @ueshin